packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -121,7 +122,12 @@ export class Grant implements cdk.IDependable {
       scope: options.resource,
     });
 
-    if (result.success) { return result; }
+    const definitelySameAccount = accountsAreSameOrUnresolved(
+      options.grantee.grantPrincipal.principalAccount,
```

Comment:
@rix0rrr can you take a look here?

Reply:
This is correct and copied from https://github.com/aws/aws-cdk/pull/5812/files, but it could have done with adding the lines of comments that explain why this is true:

```ts
    // If we added to the principal AND we're in the same account, then we're done.
    // If not, it's a different account and we must also add a trust policy on the resource.
```

Reply:
Added.